File: course_1/module_4/examples/randomized_selection.py
import random

class RandomizedSelection:
    """
    Class for finding the k-th smallest element using a randomized selection algorithm.

    Attributes:
        array (list): The input array of distinct elements.
    """

    def __init__(self, array):
        """
        Initialize the RandomizedSelection class with an array of distinct elements.
        """
        if not array:
            raise ValueError("Input array must not be empty.")
        if len(set(array)) != len(array):
            raise ValueError("Array must contain distinct elements.")
        self.array = array[:]

    def _partition(self, left, right, pivot_index):
        """
        Partition the array around the pivot element.
        """
        pivot_value = self.array[pivot_index]
        self.array[pivot_index], self.array[right] = self.array[right], self.array[pivot_index]
        store_index = left
        for i in range(left, right):
            if self.array[i] < pivot_value:
                self.array[i], self.array[store_index] = self.array[store_index], self.array[i]
                store_index += 1
        self.array[right], self.array[store_index] = self.array[store_index], self.array[right]
        return store_index

    def _randomized_select(self, left, right, k):
        """
        Recursively find the k-th smallest element in the array.
        """
        if left == right:
            return self.array[left]

        pivot_index = random.randint(left, right)
        pivot_index = self._partition(left, right, pivot_index)
        rank = pivot_index - left + 1

        if rank == k:
            return self.array[pivot_index]
        elif k < rank:
            return self._randomized_select(left, pivot_index - 1, k)
        else:
            return self._randomized_select(pivot_index + 1, right, k - rank)

    def select(self, k):
        """
        Find the k-th smallest element in the array.

        Args:
            k (int): 1-based index of the desired smallest element.

        Returns:
            int: The k-th smallest element in the array.
        """
        if k < 1 or k > len(self.array):
            raise ValueError("k is out of bounds of the array.")
        return self._randomized_select(0, len(self.array) - 1, k)


# Example usage.
if __name__ == "__main__":
    array = [10, 4, 5, 8, 6, 11, 26]
    k = 3
    selector = RandomizedSelection(array)
    result = selector.select(k)
    print(f"The {k}-th smallest element is: {result}")

File: course_1/module_4/examples/randomized_selection.go
package main

import (
	"errors"
	"fmt"
	"math/rand"
	"time"
)

// RandomizedSelection struct encapsulates the array and its methods.
type RandomizedSelection struct {
	array []int
}

// NewRandomizedSelection creates a new instance of RandomizedSelection.
func NewRandomizedSelection(array []int) (*RandomizedSelection, error) {
	if len(array) == 0 {
		return nil, errors.New("input array must not be empty")
	}
	return &RandomizedSelection{array: append([]int{}, array...)}, nil
}

// partition partitions the array around the pivot element.
func (rs *RandomizedSelection) partition(left, right, pivotIndex int) int {
	pivotValue := rs.array[pivotIndex]
	// Move pivot to the end.
	rs.array[pivotIndex], rs.array[right] = rs.array[right], rs.array[pivotIndex]
	storeIndex := left

	// Rearrange elements around the pivot.
	for i := left; i < right; i++ {
		if rs.array[i] < pivotValue {
			rs.array[i], rs.array[storeIndex] = rs.array[storeIndex], rs.array[i]
			storeIndex++
		}
	}

	// Move pivot to its final position.
	rs.array[storeIndex], rs.array[right] = rs.array[right], rs.array[storeIndex]
	return storeIndex
}

// randomizedSelect recursively finds the k-th smallest element.
func (rs *RandomizedSelection) randomizedSelect(left, right, k int) int {
	if left == right {
		return rs.array[left] // Base case: only one element.
	}

	// Choose a random pivot index.
	pivotIndex := rand.Intn(right-left+1) + left

	// Partition the array and get the pivot's final index.
	pivotIndex = rs.partition(left, right, pivotIndex)

	// Determine the rank of the pivot.
	rank := pivotIndex - left + 1

	if rank == k {
		return rs.array[pivotIndex]
	} else if k < rank {
		return rs.randomizedSelect(left, pivotIndex-1, k)
	} else {
		return rs.randomizedSelect(pivotIndex+1, right, k-rank)
	}
}

// Select finds the k-th smallest element in the array.
func (rs *RandomizedSelection) Select(k int) (int, error) {
	if k < 1 || k > len(rs.array) {
		return 0, errors.New("k is out of bounds of the array")
	}
	return rs.randomizedSelect(0, len(rs.array)-1, k), nil
}

func main() {
	// Initialize the random seed.
	rand.Seed(time.Now().UnixNano())

	array := []int{10, 4, 5, 8, 6, 11, 26}
	k := 3

	// Create an instance of RandomizedSelection.
	selector, err := NewRandomizedSelection(array)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Find the k-th smallest element.
	result, err := selector.Select(k)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("The %d-th smallest element is: %d\n", k, result)
}
